src/services/semver.ts

```diff
@@ -0,0 +1,57 @@
+/* @internal */
+namespace ts {
+    function stringToInt(str: string): number {
+        const n = parseInt(str, 10);
+        if (isNaN(n)) {
+            throw new Error(`Error in parseInt(${JSON.stringify(str)})`);
+        }
+        return n;
+    }
+
+    export class Semver {
+        static parse(semver: string): Semver {
+            const isPrerelease = /^(.*)-next.\d+/.test(semver);
+            const result = Semver.tryParse(semver, isPrerelease);
+            if (!result) {
+                throw new Error(`Unexpected semver: ${semver} (isPrerelease: ${isPrerelease})`);
+            }
+            return result;
+        }
+
+        static fromRaw({ major, minor, patch, isPrerelease }: Semver): Semver {
+            return new Semver(major, minor, patch, isPrerelease);
+        }
+
+        // This must parse the output of `versionString`.
+        private static tryParse(semver: string, isPrerelease: boolean): Semver | undefined {
+            // Per the semver spec <http://semver.org/#spec-item-2>:
+            // "A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, and MUST NOT contain leading zeroes."
+            const rgx = isPrerelease ? /^(\d+)\.(\d+)\.0-next.(\d+)$/ : /^(\d+)\.(\d+)\.(\d+)$/;

```

Comment:
for these regexp's i would move them to locals to avoid parsing/creating a new one on every invocation.

Reply:
Moved.